send/generic_sender.py

```diff
@@ -0,0 +1,200 @@
+		service_files_for_destination = service_files_dir + "/_destination/all"
+
+	if os.path.isdir(service_files_for_destination):
+		tar_command.extend(["-C", service_files_for_destination, "."])
+
+	return tar_command
+
+
+def prepare_url_transport_command(temp, content_type: str = "application/x-tar", method: str = "PUT"):
+	"""
+	Prepares base transport command for CURL connection.
+	If available, adds common perun certificate.
+	:param temp:
+	:param content_type: for example 'application/json'
+	:return: transport command
+	"""
+	transport_command = ["curl"]
+	# add certificate to the curl if cert file and key file exist, and they are readable
+	if os.access(send_lib.PERUN_CERT, os.R_OK) and os.access(send_lib.PERUN_KEY, os.R_OK) and os.access(
+		send_lib.PERUN_CHAIN, os.R_OK):
+		transport_command.extend(
+			["--cert", send_lib.PERUN_CERT, "--key", send_lib.PERUN_KEY, "--cacert", send_lib.PERUN_CHAIN])
+	# add standard CURL params
+	transport_command.extend(["-i", "-H", "Content-Type:" + content_type, "-w", "%{http_code}", "--show-error",
+							  "--silent", "-o", temp.name, "-X", method, "--data-binary", "@-"])
+	return transport_command
+
+
+def prepare_ssh_transport_command():
+	"""
+	Prepares base transport command for SSH connection.
+	:return: transport command
+	"""
+	return ["ssh", "-o", "PasswordAuthentication=no", "-o", "StrictHostKeyChecking=no", "-o",
+			"GSSAPIAuthentication=no", "-o", "GSSAPIKeyExchange=no", "-o", "ConnectTimeout=5"]
+
+
+if __name__ == "__main__":
+	send_lib.check_input_fields(sys.argv)
+	facility_name = sys.argv[1]
+	destination = sys.argv[2]
+	if len(sys.argv) == 3:
+		# if there is no destination type, use default 'host'
+		destination_type = send_lib.DESTINATION_TYPE_HOST
+	else:
+		destination_type = sys.argv[3]
+		if destination_type == send_lib.DESTINATION_TYPE_EMAIL or destination_type == send_lib.DESTINATION_TYPE_SERVICE_SPECIFIC:
+			print("Destination type " + destination_type + " is not supported yet.", file=sys.stderr)
+			exit(1)
+
+	send_lib.check_destination_format(destination, destination_type)
+	service_name = send_lib.get_global_service_name()
+
+	# choose transport command, only url type has different transport command at this moment
+	transport_command = send_lib.load_custom_transport_command(service_name)
+	if transport_command is None and destination_type == send_lib.DESTINATION_TYPE_URL:
+		# FIXME - test this use-case
+		temp = tempfile.NamedTemporaryFile(mode="w+")
+		# errors will be saved to temp file
+		transport_command = prepare_url_transport_command(temp)
+	elif transport_command is None:
+		transport_command = prepare_ssh_transport_command()
+
+	if destination_type == send_lib.DESTINATION_TYPE_USER_HOST_WINDOWS_PROXY:
+		windows_proxy = send_lib.get_windows_proxy()
+		send_lib.check_windows_proxy_format(windows_proxy)
+		host, hostname, port = send_lib.prepare_destination(destination, destination_type, windows_proxy)
+		transport_command.extend(["-p", port])
```

Comment:
Wouldn't a separate condition that adds the port option if 'port' is defined be better for extensibility (if we added more destination types with a custom port)? Then we could also move the windows proxy logic to send_lib (since all we do here is call functions from send_lib anyway). Also, 'DESTINATION_TYPE_USER_HOST_PORT' uses a custom port and is not accounted for here. Not sure whether this destination type is actually used, though.

Reply:
good idea! i think i unintentionally changed the usage in the generic_sender and put it to condition for USER_HOST_WINDOWS_PROXY instead of for USER_HOST_PORT